Add support for SX126x transceivers

andysan opened the pull request:
This pull request adds support for the SX1261 and SX1262 LoRa modems.

Due to a significant overlap between the SX1276 driver and these drives, some common functionality has been factored out into shared source files.

The current version has the following known limitations:

- The radio switch can only be controlled by DIO2. RF switches controlled by GPIOs are currently not supported (mostly due to a limitation in the HAL).
- A TCXO, if present, must be controlled by DIO3 or always on. GPIO contol is currently not possible.

Testing:

- Broadcasting of a continuous wave works.
- Sending packets has been tested and is working.
- Packet reception is working.
- Tested with @Mani-Sadhasivam's LoRaWAN patches. Able to join TTN using OTAA and send packets.

The above was verified in a lab setup with an SX126X connected to an NRF52840 dev board and an SDR through a set of RF splitters and and attenuators. TTN testing was done over the air.

Dependencies:
- ~https://github.com/zephyrproject-rtos/loramac-node/pull/4~
- ~#24577 for CW API changes~
- ~#25040 for the SX1276 refactor as a separate PR.~

andysan committed:
drivers: lora: Add SX126x LoRa radio support

Add support for the SX126x series of LoRa radios using the
LoRaMAC-Node HAL.

This driver currently makes the following assumptions:

  * DIO1 is used as an interrupt line.

  * There is an RF switch selecting between the TX and RX ports and
    that switch is controlled by DIO2.

  * There is either no TCXO or the TCXO is controlled by DIO3.

Specifically, the limitations above mean that modules that use GPIOs
to control the RF switch are currently not supported. Support for such
modules would need changes to the LoRaMAC-Node code.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

galak reviewed (commented):
I'd expected a west.yml update if the module is change, or support for SX126x already in loramac-node?

andysan commented:
> I'd expected a west.yml update if the module is change, or support for SX126x already in loramac-node?

Yep, that's correct. The pull request can be found here: https://github.com/zephyrproject-rtos/loramac-node/pull/4

I'll update this RFC when that has been merged.

carlescufi commented:
> > I'd expected a west.yml update if the module is change, or support for SX126x already in loramac-node?
> 
> Yep, that's correct. The pull request can be found here: [zephyrproject-rtos/loramac-node#4](https://github.com/zephyrproject-rtos/loramac-node/pull/4)
> 
> I'll update this RFC when that has been merged.

No, please update it now with `pull/4/head` until that is merged. See here:
https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules

andysan commented:
> pull/4/head



> > > I'd expected a west.yml update if the module is change, or support for SX126x already in loramac-node?
> > 
> > 
> > Yep, that's correct. The pull request can be found here: [zephyrproject-rtos/loramac-node#4](https://github.com/zephyrproject-rtos/loramac-node/pull/4)
> > I'll update this RFC when that has been merged.
> 
> No, please update it now with `pull/4/head` until that is merged. See here:
> https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules

I missed that part of the contribution guide, sorry. :facepalm: I'll update the pull request right away.

carlescufi commented:
> > pull/4/head
> 
> > > > I'd expected a west.yml update if the module is change, or support for SX126x already in loramac-node?
> > > 
> > > 
> > > Yep, that's correct. The pull request can be found here: [zephyrproject-rtos/loramac-node#4](https://github.com/zephyrproject-rtos/loramac-node/pull/4)
> > > I'll update this RFC when that has been merged.
> > 
> > 
> > No, please update it now with `pull/4/head` until that is merged. See here:
> > https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules
> 
> I missed that part of the contribution guide, sorry. 🤦 I'll update the pull request right away.

No worries, thank you for your PR.

carlescufi commented:
@andysan the reason I asked you is that there is no CI in the module repos, so we have to rely on this repository's CI to make sure the module's PR doesn't break anything.

gmarull commented on drivers/lora/sx12xx_common.c:
```diff
@@ -0,0 +1,213 @@
+
+int sx12xx_lora_init(struct device *dev)
+{
+	dev_data.counter = device_get_binding(DT_RTC_0_NAME);
```
Shouldn't new DT API be used here? Still, I think it would be better to specify the RTC instance via a DT binding. cc: @galak @mbolivar-nordic

andysan replied:
Short answer: Yes. The slightly longer answer is that @Mani-Sadhasivam is rewriting this code in his LoRaWAN pull request and this commit only re-arranges the existing code.

mbolivar-nordic replied:
I didn't look at this closely, but if it's just moving code around it seems fine for now. Though the old API is going to be unavailable for in-tree code by release time.

andysan replied:
This has been fix by @Mani-Sadhasivam on master. Rebased version will reflect the change.

mniestroj commented on drivers/lora/sx126x.c:
```diff
@@ -0,0 +1,480 @@
+		}
+	};
+
+	const struct spi_buf rx_buf[] = {
```
You can use the same array of buffers for tx and rx. This would allow to drop `req_rx` argument. Maybe the same could be done with `data_rx`, because there is no case when `data_tx` and `data_rx` are passed simultaneously. In spi_nor.c and sx1276.c there is a `is_write` argument instead.

andysan replied:
I tried that strategy and it didn't seem to work on my nrf52 board. I suspect this is similar to an issues discussed on Slack when the sx1276 is used on that platform. I didn't debug it properly, but I think the DMA operation in the SPI peripheral gets confused when the same pointer is used in both the input and output register.

mniestroj commented on drivers/lora/sx126x.c:
```diff
@@ -0,0 +1,480 @@
+	dev_data.spi_cfg.operation = SPI_WORD_SET(8) | SPI_TRANSFER_MSB;
+	dev_data.spi_cfg.frequency = DT_INST_PROP(0, spi_max_frequency);
+	dev_data.spi_cfg.slave = DT_INST_REG_ADDR(0);
+	dev_data.spi_cfg.cs = &spi_cs;
```
Make CS optional, same as in #24470.

andysan replied:
Ack

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -439,6 +439,13 @@ static int sx1276_lora_config(struct device *dev,
 	return 0;
 }
 
+int sx1276_lora_test_cw(struct device *dev, u32_t frequency,
```
`static` (commented about it in another PR as well)

mniestroj commented on drivers/lora/Kconfig.semtech:
```diff
@@ -25,6 +25,12 @@ config LORA_SX1276
 
+config LORA_SX126X
+	bool "Semtech SX126x driver"
+	select HAS_SEMTECH_SX126X
```
I wonder what is the use of HAS_SEMTECH_SX126X. It is selected, but not used anywhere.

andysan replied:
It's used in the LoRaMAC-node repo to include the SX126X driver in the build. See https://github.com/zephyrproject-rtos/loramac-node/pull/4.

mniestroj commented on drivers/lora/Kconfig.semtech:
```diff
@@ -4,16 +4,29 @@
+
+choice
+	prompt "LoRa Radio chipset"
+	default LORA_SX1276
```
I think that we should have some automatic selection of the right driver based on dt compatible. With current implementation user will need to select LORA_SX126X explicitly, in addition to configured device-tree node.

andysan replied:
I'm not sure if I know how to implement that in a good way. We would need to build different source files based on DT properties, which I don't think is possible. We could of course use some weird include hack that pulls in a radio implementation depending on the DT, but that seems quite scary.

mbolivar-nordic replied:
> We would need to build different source files based on DT properties, which I don't think is possible.

It is. Please look in scripts/kconfig/kconfigfunctions.py. For example dt_compat_enabled() will let you detect if a devicetree node of a particular compatible is turned on. You can search for uses in the rest of the code for example uses.

andysan replied:
Cool. I'll make sure to use it in that case. Thanks for the pointer!

mbolivar-nordic replied:
NP. Note that since you can only access Kconfig values in CMake, and not dt properties, the workaround for now is to do something like

``
config MY_DRIVER_WHATEVER
    default y if $(dt_something_something,...)
```

then build your driver if CONFIG_MY_DRIVER_WHATEVER=y in CMake.

andysan replied:
I'm not super familiar with Kconfig, but I think I implemented this the "right" way. I ended up adding a choice of chip types and defaulting to the one in the DT.

mniestroj reviewed (commented):
Overall looks okay. I would like to test that on my custom board as soon as you rebase on current master branch.

andysan committed:
drivers: lora: Prepare Kconfig for additional radios

Add a new configuration option, LORA_SX12XX, that is shared for all of
the LoRaMAC-node-based radio drivers. By default, the appropriate
driver for the LoRa radio in the device tree is included is included
in the build.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

andysan committed:
dts: bindings: Add bindings for Semtech SX126x

Add device tree bindings for the Semtech SX1261 and SX1262
radios. These will be used by Zephyr's LoRa drivers.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

mbolivar-nordic commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -0,0 +1,62 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+DT_COMPAT_SX1276 = semtech,sx1276
```
Please add SEMTECH_ before each, for consistency with the rest of the Kconfig code.

You may also want to consider using `:=` instead of `=`. The former expands once, the latter each time the variable is read (kind of like Make's two types of variables). Since you have no need for re-expansion it would be clearer to use `:=`.

Like this:

```
DT_COMPAT_NORDIC_NRF_SPI  := nordic,nrf-spi
```

But instead:

```
DT_COMPAT_SEMTECH_SX1276  := semtech,sx1276
```

etc.

andysan replied:
Ack

mbolivar-nordic commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -0,0 +1,62 @@
+
+menuconfig LORA_SX12XX
+	bool "Semtech SX-series driver"
+	select HAS_SEMTECH_RADIO_DRIVERS
```
Why are these selects necessary? Isn't `#ifdef CONFIG_LORA_SX12XX` enough in software?

Usually, `HAS_FOO` macros are used to signal the presence of some hardware, or when an arch supports a feature, or something like that. This seems a bit strange.

andysan replied:
They are used to control the build of the LoRaMAC-Node library. The `CONFIG_HAS_LORAMAC` macro includes the MAC and the `CONFIG_HAS_SEMTECH_SX*` macros select which radio to build in the library.

Not sure what the rationale was since these were used by the original SX1276 driver.

mbolivar-nordic replied:
If this is just moving existing options around, OK, but this looks like unconventional usage. Thanks.

mbolivar-nordic commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -0,0 +1,62 @@
+if LORA_SX12XX
+
+choice
+	prompt "LoRa Radio chipset"
```
Is this really a choice? Is it possible to have two different LoRa radios in one system, one with an SX1276 and another with an SX1261? If so, I think these would be best expressed as separate `bool` options, each of which enables support for building an individual driver. Kind of like the example in https://github.com/zephyrproject-rtos/zephyr/pull/24616#discussion_r420311757.

andysan replied:
The LoRaMac-node library assumes that there is only one LoRa radio in the system. For that reason, it has been designed using global struct to bind the radio and the MAC. There is no easy way to work around that without extensive hacking of the library.

For boards that have both (I was in fact testing with such a board for a little while), the user would have to select which radio to use at build time.

mbolivar-nordic commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -0,0 +1,62 @@
+
+config LORA_SX1276
+	bool "Semtech SX1276 driver"
+	select HAS_SEMTECH_SX1276
```
These selects also seem strange. Why are they necessary?

In software, you can just test if any device with the right compatible is enabled using the devicetree API.

andysan replied:
See comment above about LoRaMAC-node.

mbolivar-nordic replied:
OK

mbolivar-nordic commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -0,0 +1,62 @@
+endchoice
+
+choice
+	prompt "SX1276 PA Output pin"
```
This looks like it might belong in the devicetree binding as a property, instead of in Kconfig, no?

andysan replied:
Simple answer -- yes. @mniestroj has created a pull request (#24169) to fix that. This pull request doesn't change the SX1276 driver other than factoring out common code to facilitate reuse.

mbolivar-nordic replied:
Great, thanks

carlescufi commented:
@Mani-Sadhasivam could you review this PR please?

Mani-Sadhasivam commented:
> @Mani-Sadhasivam could you review this PR please?

@carlescufi Sorry, a bit occupied. Will do it in couple of days. Are you planning to get this into 2.3?

andysan commented:
> > @Mani-Sadhasivam could you review this PR please?
> 
> 
> 
> @carlescufi Sorry, a bit occupied. Will do it in couple of days. Are you planning to get this into 2.3?

There is no particular rush from my side to merge the driver itself, but it’s be nice to merge the refactor of the 1276 since that’s causing merge conflicts on my work branches. Happy to split the PR if that speeds up reviews.

Mani-Sadhasivam commented:
> There is no particular rush from my side to merge the driver itself, but it’s be nice to merge the refactor of the 1276 since that’s causing merge conflicts on my work branches. Happy to split the PR if that speeds up reviews.

Please do so!

andysan commented:
> > There is no particular rush from my side to merge the driver itself, but it’s be nice to merge the refactor of the 1276 since that’s causing merge conflicts on my work branches. Happy to split the PR if that speeds up reviews.
> 
> Please do so!

I have created #25040 with the first 3 commits. Other than the first commit (the CW API change from the shell PR), there is no significant new code in those commits.

Mani-Sadhasivam commented:
I'm keeping this PR in backlog until the dependent PRs are merged/reviewed.

carlescufi commented:
Please update west.yml with a SHA

carlescufi commented:
@Mani-Sadhasivam can you please approve if happy?

Mani-Sadhasivam commented:
> @Mani-Sadhasivam can you please approve if happy?

On my TODO list. Will do it in couple of days.

andysan commented:
Further testing with the LoRaWAN patches suggests that something weird is happening. For some reason, I can’t receive downlinks correctly. Please keep this as DNM until I have investigated that.

andysan commented:
> Further testing with the LoRaWAN patches suggests that something weird is happening. For some reason, I can’t receive downlinks correctly. Please keep this as DNM until I have investigated that.

Turns out this was caused by the LoRaWAN stack performing SPI transactions from an interrupt context. As far as I can tell, the weird errors I was experiencing weren't caused by this driver.

mniestroj commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -0,0 +1,34 @@
+	prompt "LoRa Radio chipset"
+	default LORA_SX1276 if $(dt_compat_enabled,$(DT_COMPAT_SEMTECH_SX1276))
+	help
+	  Antenna connection type.
```
This description is wrong, choice is about "LoRa Radio chipset".

andysan replied:
Ack

mniestroj commented on drivers/lora/sx126x.c:
```diff
@@ -0,0 +1,479 @@
+	DT_INST_GPIO_FLAGS(0, antenna_enable_gpios)
+
+#if DT_INST_NODE_HAS_PROP(0, dio2_tx_enable)
+#define DIO2_TX_ENABLE DT_INST_PROP(0, dio2_tx_enable)
```
Boolean values (such as `DT_INST_PROP(0, dio2_tx_enable)`) are always set to 0  or 1, so there is no need to check with `DT_INST_NODE_HAS_PROP()`. @mbolivar-nordic can you confirm that?

andysan replied:
Ack. Tested without the conditional enable and it seems to behave as expected.

mniestroj commented on dts/bindings/lora/semtech,sx126x-base.yaml:
```diff
@@ -0,0 +1,56 @@
+      description: |
+        GPIO connected to the sensor BUSY signal.
+
+    irq-gpios:
```
I would have a slight preference for dio1-gpios, as DIO1 is what we see usually on schematics. `busy-gpios` can be configured as interrupt as well in the future, so we could get rid of busy (with k_sleep) loop some day.

andysan replied:
That makes sense. I have updated the bindings.

mniestroj replied:
Bindings are updated, but not in the right commit :)

andysan replied:
Sigh, I thought I made sure to apply the fixups to the right commits. I’ll post an update tonight.

mniestroj reviewed (changes requested):
Tested LoRa shell with custom board and sx1262. Just some minor things to improve.

mniestroj commented on dts/bindings/lora/semtech,sx126x-base.yaml:
```diff
@@ -0,0 +1,56 @@
+      type: phandle-array
+      required: true
+      description: |
+        GPIO connected to the sensor NRST signal.
```
s/sensor/modem/ ?

andysan replied:
ack

mniestroj commented on dts/bindings/lora/semtech,sx126x-base.yaml:
```diff
@@ -0,0 +1,56 @@
+      type: phandle-array
+      required: true
+      description: |
+        GPIO connected to the sensor BUSY signal.
```
s/sensor/modem/ ?

andysan replied:
Weird. Must have been a copy-and-paste error. It should be fixed now.

mniestroj reviewed (commented):
I just noticed minor documentation issues in DT bindings.

mniestroj reviewed (approved):
LGTM

Mani-Sadhasivam commented on drivers/lora/sx126x.c:
```diff
@@ -0,0 +1,475 @@
+	gpio_init_callback(&dev_data.dio1_irq_callback,
+			   sx126x_dio1_irq_callback, BIT(GPIO_DIO1_PIN));
+	if (gpio_add_callback(dev_data.dio1, &dev_data.dio1_irq_callback) < 0) {
+		LOG_ERR("Could not set GPIO callback for IRQ line.");
```
Showing the IRQ line number would be helpful here.

andysan replied:
I have rephrased this as "Could not set GPIO callback for DIO1 interrupt" I think that should be a bit clearer.

Mani-Sadhasivam commented on drivers/lora/sx126x.c:
```diff
@@ -0,0 +1,475 @@
+{
+	LOG_DBG("Processing IRQ");
+	if (!dev_data.radio_dio_irq) {
+		LOG_WRN("Radio IRQ before setting radio_dio_irq.");
```
Can we say, `DIO1 IRQ callback not defined` instead?

andysan replied:
Rephrased as "DIO1 interrupt without valid HAL IRQ callback.".

Mani-Sadhasivam reviewed (approved):
LGTM!

andysan commented:
@carlescufi This should be ready to be merged now. The test issue seems unrelated (`tests/subsys/logging/log_immediate/logging.log_immediate` failing on ARC).

holmes81 commented:
Hi,Can please help to check this:
https://github.com/zephyrproject-rtos/zephyr/issues/27915

It can't work property with nrf52840dk,I don't know why.There is no data from MISO,
Thanks a lot.

holmes81 commented:
Or can please share your firmware bin file,then I can have a test.
